

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>sonia package &mdash; sonia 0.0.26 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="License" href="license.html" />
    <link rel="prev" title="Documentation" href="modules.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> sonia
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="readme.html">Directory architecture</a></li>
<li class="toctree-l1"><a class="reference internal" href="commands.html">Command line console scripts</a></li>
<li class="toctree-l1"><a class="reference internal" href="python_usage.html">Sonia modules in a Python script</a></li>
<li class="toctree-l1"><a class="reference internal" href="sonia_tutorial.html">Sonia Example</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">Documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">sonia package</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-sonia.evaluate_model">sonia.evaluate_model module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-sonia.plotting">sonia.plotting module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-sonia.sequence_generation">sonia.sequence_generation module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-sonia.sonia">sonia.sonia module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-sonia.sonia_leftpos_rightpos">sonia.sonia_leftpos_rightpos module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-sonia.sonia_length_pos">sonia.sonia_length_pos module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-sonia.sonia_vjl">sonia.sonia_vjl module</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Need Help?</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">sonia</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="modules.html">Documentation</a> &raquo;</li>
        
      <li>sonia package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/sonia.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="sonia-package">
<h1>sonia package<a class="headerlink" href="#sonia-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-sonia.evaluate_model">
<span id="sonia-evaluate-model-module"></span><h2>sonia.evaluate_model module<a class="headerlink" href="#module-sonia.evaluate_model" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="sonia.evaluate_model.EvaluateModel">
<em class="property">class </em><code class="sig-prename descclassname">sonia.evaluate_model.</code><code class="sig-name descname">EvaluateModel</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sonia_model</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">include_genes</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">processes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">custom_olga_model</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.evaluate_model.EvaluateModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class used to evaluate sequences with the sonia model: Ppost=Q*Pgen</p>
<dl class="simple">
<dt>sonia_model: object</dt><dd><p>Sonia model. Loaded previously, do not put the path.</p>
</dd>
<dt>include_genes: bool</dt><dd><p>Conditioning on gene usage for pgen/ppost evaluation. Default: True</p>
</dd>
<dt>processes: int</dt><dd><p>Number of processes to use to infer pgen. Default: all.</p>
</dd>
<dt>custom_olga_model: object</dt><dd><p>Optional: already loaded custom generation_probability olga model.</p>
</dd>
</dl>
<dl class="simple">
<dt>evaluate_seqs(seqs=[])</dt><dd><p>Returns Q, pgen and ppost of a list of sequences.</p>
</dd>
<dt>evaluate_selection_factors(seqs=[])</dt><dd><p>Returns normalised selection factor Q (Ppost=Q*Pgen) of a list of sequences (faster than evaluate_seqs because it does not compute pgen and ppost)</p>
</dd>
</dl>
<dl class="py method">
<dt id="sonia.evaluate_model.EvaluateModel.compute_all_pgens">
<code class="sig-name descname">compute_all_pgens</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seqs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.evaluate_model.EvaluateModel.compute_all_pgens" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute Pgen of sequences using OLGA in parallel</p>
<dl class="simple">
<dt>seqs: list</dt><dd><p>list of sequences to evaluate.</p>
</dd>
</dl>
<dl class="simple">
<dt>pgens: array</dt><dd><p>generation probabilities of the sequences.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.evaluate_model.EvaluateModel.compute_joint_marginals">
<code class="sig-name descname">compute_joint_marginals</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sonia.evaluate_model.EvaluateModel.compute_joint_marginals" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes joint marginals for all.</p>
<dl class="simple">
<dt>gen_marginals_two: array</dt><dd><p>matrix (i,j) of joint marginals for pre-selection distribution</p>
</dd>
<dt>data_marginals_two: array</dt><dd><p>matrix (i,j) of joint marginals for data</p>
</dd>
<dt>model_marginals_two: array</dt><dd><p>matrix (i,j) of joint marginals for post-selection distribution</p>
</dd>
<dt>gen_marginals_two_independent: array</dt><dd><p>matrix (i,j) of independent joint marginals for pre-selection distribution</p>
</dd>
<dt>data_marginals_two_independent: array</dt><dd><p>matrix (i,j) of joint marginals for pre-selection distribution</p>
</dd>
<dt>model_marginals_two_independent: array</dt><dd><p>matrix (i,j) of joint marginals for pre-selection distribution</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.evaluate_model.EvaluateModel.evaluate_selection_factors">
<code class="sig-name descname">evaluate_selection_factors</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seqs</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.evaluate_model.EvaluateModel.evaluate_selection_factors" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns normalised selection factor Q (of Ppost=Q*Pgen) of list of sequences (faster than evaluate_seqs because it does not compute pgen and ppost)</p>
<dl class="simple">
<dt>seqs: list</dt><dd><p>list of sequences to evaluate</p>
</dd>
</dl>
<dl class="simple">
<dt>Q: array</dt><dd><p>selection factor Q (of Ppost=Q*Pgen) of the sequences</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.evaluate_model.EvaluateModel.evaluate_seqs">
<code class="sig-name descname">evaluate_seqs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seqs</span><span class="o">=</span><span class="default_value">[]</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.evaluate_model.EvaluateModel.evaluate_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns selection factors, pgen and pposts of sequences.</p>
<dl class="simple">
<dt>seqs: list</dt><dd><p>list of sequences to evaluate</p>
</dd>
</dl>
<dl class="simple">
<dt>Q: array</dt><dd><p>selection factor Q (of Ppost=Q*Pgen) of the sequences</p>
</dd>
<dt>pgens: array</dt><dd><p>pgen of the sequences</p>
</dd>
<dt>pposts: array</dt><dd><p>ppost of the sequences</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.evaluate_model.EvaluateModel.joint_marginals">
<code class="sig-name descname">joint_marginals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">features</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">seq_model_features</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">seqs</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">use_flat_distribution</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.evaluate_model.EvaluateModel.joint_marginals" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Returns joint marginals P(i,j) with i and j features of sonia (l3, aA6, etc..), index of features attribute is preserved.</dt><dd><p>Matrix is upper-triangular.</p>
</dd>
</dl>
<dl class="simple">
<dt>features: list</dt><dd><p>custom feature list</p>
</dd>
<dt>seq_model_features: list</dt><dd><p>encoded sequences</p>
</dd>
<dt>seqs: list</dt><dd><p>seqs to encode.</p>
</dd>
<dt>use_flat_distribution: bool</dt><dd><p>for data and generated seqs is True, for model is False (weights with Q)</p>
</dd>
</dl>
<dl class="simple">
<dt>joint_marginals: array</dt><dd><p>matrix (i,j) of joint marginals</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.evaluate_model.EvaluateModel.joint_marginals_independent">
<code class="sig-name descname">joint_marginals_independent</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">marginals</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.evaluate_model.EvaluateModel.joint_marginals_independent" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns independent joint marginals P(i,j)=P(i)*P(j) with i and j features of sonia (l3, aA6, etc..), index of features attribute is preserved.
Matrix is upper-triangular.</p>
<dl class="simple">
<dt>marginals: list</dt><dd><p>marginals.</p>
</dd>
</dl>
<dl class="simple">
<dt>joint_marginals: array</dt><dd><p>matrix (i,j) of joint marginals</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-sonia.plotting">
<span id="sonia-plotting-module"></span><h2>sonia.plotting module<a class="headerlink" href="#module-sonia.plotting" title="Permalink to this headline">¶</a></h2>
<p>Command line script to generate sequences.</p>
<p>Copyright (C) 2020 Isacchini Giulio</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="https://www.gnu.org/licenses/">https://www.gnu.org/licenses/</a>&gt;.</p>
<dl class="py class">
<dt id="sonia.plotting.Plotter">
<em class="property">class </em><code class="sig-prename descclassname">sonia.plotting.</code><code class="sig-name descname">Plotter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sonia_model</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.plotting.Plotter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class used to plot stuff</p>
<dl class="simple">
<dt>sonia_model: object</dt><dd><p>Sonia model. No path.</p>
</dd>
</dl>
<dl class="simple">
<dt>plot_model_learning(save_name = None)</dt><dd><p>Plots L1 convergence curve and marginal scatter.</p>
</dd>
<dt>plot_pgen(pgen_data=[],pgen_gen=[],pgen_model=[],n_bins=100)</dt><dd><p>Histogram plot of pgen. You need to evalute them first.</p>
</dd>
<dt>plot_ppost(ppost_data=[],ppost_gen=[],pppst_model=[],n_bins=100)</dt><dd><p>Histogram plot of ppost. You need to evalute them first.</p>
</dd>
<dt>plot_model_parameters(low_freq_mask = 0.0)</dt><dd><p>For LengthPos model only. Plot the model parameters using plot_onepoint_values</p>
</dd>
<dt>plot_marginals_length_corrected(min_L = 8, max_L = 16, log_scale = True)</dt><dd><p>For LengthPos model only. Plot length normalized marginals.</p>
</dd>
<dt>plot_vjl(save_name = None)</dt><dd><p>Plots marginals of V gene, J gene and cdr3 length</p>
</dd>
<dt>plot_logQ(save_name=None)</dt><dd><p>Plots logQ of data and generated sequences</p>
</dd>
<dt>plot_ratioQ(self,save_name=None)</dt><dd><p>Plots the ratio of P(Q) in data and pre-selected pool. Useful for model validation.</p>
</dd>
</dl>
<dl class="py method">
<dt id="sonia.plotting.Plotter.norm_marginals">
<code class="sig-name descname">norm_marginals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">marg</span></em>, <em class="sig-param"><span class="n">min_L</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">max_L</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.plotting.Plotter.norm_marginals" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>renormalizing the marginals accourding to length, so the sum of the marginals over all amino acid </dt><dd><p>for one position/length combination will be 1 (and not the fraction of CDR3s of this length)</p>
</dd>
</dl>
<dl class="simple">
<dt>marg<span class="classifier">ndarray</span></dt><dd><p>the marginal to renormalize</p>
</dd>
<dt>min_L<span class="classifier">int</span></dt><dd><p>Minimum length CDR3 sequence, if not given taken from class attribute</p>
</dd>
<dt>max_L<span class="classifier">int</span></dt><dd><p>Maximum length CDR3 sequence, if not given taken from class attribute</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.plotting.Plotter.plot_logQ">
<code class="sig-name descname">plot_logQ</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">save_name</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.plotting.Plotter.plot_logQ" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots logQ of data and generated sequences</p>
<dl class="simple">
<dt>save_name<span class="classifier">str or None</span></dt><dd><p>File name to save output figure. If None (default) does not save.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.plotting.Plotter.plot_marginals_length_corrected">
<code class="sig-name descname">plot_marginals_length_corrected</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">min_L</span><span class="o">=</span><span class="default_value">8</span></em>, <em class="sig-param"><span class="n">max_L</span><span class="o">=</span><span class="default_value">16</span></em>, <em class="sig-param"><span class="n">log_scale</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.plotting.Plotter.plot_marginals_length_corrected" title="Permalink to this definition">¶</a></dt>
<dd><p>plot length normalized marginals using plot_onepoint_values</p>
<dl class="simple">
<dt>min_L<span class="classifier">int</span></dt><dd><p>Minimum length CDR3 sequence, if not given taken from class attribute</p>
</dd>
<dt>max_L<span class="classifier">int</span></dt><dd><p>Maximum length CDR3 sequence, if not given taken from class attribute</p>
</dd>
<dt>log_scale<span class="classifier">bool</span></dt><dd><p>if True (default) plots marginals on a log scale</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.plotting.Plotter.plot_model_learning">
<code class="sig-name descname">plot_model_learning</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">save_name</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.plotting.Plotter.plot_model_learning" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots L1 convergence curve and marginal scatter.</p>
<dl class="simple">
<dt>save_name<span class="classifier">str or None</span></dt><dd><p>File name to save output figure. If None (default) does not save.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.plotting.Plotter.plot_model_parameters">
<code class="sig-name descname">plot_model_parameters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">low_freq_mask</span><span class="o">=</span><span class="default_value">0.0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.plotting.Plotter.plot_model_parameters" title="Permalink to this definition">¶</a></dt>
<dd><p>plot the model parameters using plot_onepoint_values</p>
<dl class="simple">
<dt>low_freq_mask<span class="classifier">float</span></dt><dd><p>threshold on the marginals, anything lower would be grayed out</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.plotting.Plotter.plot_onepoint_values">
<code class="sig-name descname">plot_onepoint_values</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">onepoint</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">onepoint_dict</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">min_L</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">max_L</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">min_val</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">max_value</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">title</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">cmap</span><span class="o">=</span><span class="default_value">'seismic'</span></em>, <em class="sig-param"><span class="n">bad_color</span><span class="o">=</span><span class="default_value">'black'</span></em>, <em class="sig-param"><span class="n">aa_color</span><span class="o">=</span><span class="default_value">'white'</span></em>, <em class="sig-param"><span class="n">marginals</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.plotting.Plotter.plot_onepoint_values" title="Permalink to this definition">¶</a></dt>
<dd><p>plot a function of aa, length and position from left, one heatplot per aa</p>
<dl class="simple">
<dt>onepoint<span class="classifier">ndarray</span></dt><dd><p>array containting one-point values to plot, in the same shape as self.features, 
expected unless onepoint_dict is given</p>
</dd>
<dt>onepoint_dict<span class="classifier">dict</span></dt><dd><p>dict of the one-point values to plot, keyed by the feature tuples such as (l12,aA8)</p>
</dd>
<dt>min_L<span class="classifier">int</span></dt><dd><p>Minimum length CDR3 sequence</p>
</dd>
<dt>max_L<span class="classifier">int</span></dt><dd><p>Maximum length CDR3 sequence</p>
</dd>
<dt>min_val<span class="classifier">float</span></dt><dd><p>minimum value to plot</p>
</dd>
<dt>max_val<span class="classifier">float</span></dt><dd><p>maximum value to plot</p>
</dd>
<dt>title<span class="classifier">string</span></dt><dd><p>title of plot to display</p>
</dd>
<dt>cmap<span class="classifier">colormap </span></dt><dd><p>colormap to use for the heatplots</p>
</dd>
<dt>bad_color<span class="classifier">string</span></dt><dd><p>color to use for nan values - used primarly for cells where position is larger than length</p>
</dd>
<dt>aa_color<span class="classifier">string</span></dt><dd><p>color to use for amino acid names for each heatplot displayed on the bad_color background</p>
</dd>
<dt>marginals<span class="classifier">bool</span></dt><dd><p>if true, indicates marginals are to be plotted and this sets cmap, bad_color and aa_color</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.plotting.Plotter.plot_prob">
<code class="sig-name descname">plot_prob</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">gen</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">model</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">n_bins</span><span class="o">=</span><span class="default_value">30</span></em>, <em class="sig-param"><span class="n">save_name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">bin_min</span><span class="o">=</span><span class="default_value">- 20</span></em>, <em class="sig-param"><span class="n">bin_max</span><span class="o">=</span><span class="default_value">- 5</span></em>, <em class="sig-param"><span class="n">ptype</span><span class="o">=</span><span class="default_value">'P_{pre}'</span></em>, <em class="sig-param"><span class="n">figsize</span><span class="o">=</span><span class="default_value">6, 4</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.plotting.Plotter.plot_prob" title="Permalink to this definition">¶</a></dt>
<dd><p>Histogram plot of Pgen</p>
<dl class="simple">
<dt>n_bins: int</dt><dd><p>number of bins of the histogram</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.plotting.Plotter.plot_ratioQ">
<code class="sig-name descname">plot_ratioQ</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">save_name</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.plotting.Plotter.plot_ratioQ" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the ratio of P(Q) in data and pre-selected pool. Useful for model validation.</p>
<dl class="simple">
<dt>save_name<span class="classifier">str or None</span></dt><dd><p>File name to save output figure. If None (default) does not save.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.plotting.Plotter.plot_vjl">
<code class="sig-name descname">plot_vjl</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">save_name</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.plotting.Plotter.plot_vjl" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots marginals of V gene, J gene and cdr3 length</p>
<dl class="simple">
<dt>save_name<span class="classifier">str or None</span></dt><dd><p>File name to save output figure. If None (default) does not save.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-sonia.sequence_generation">
<span id="sonia-sequence-generation-module"></span><h2>sonia.sequence_generation module<a class="headerlink" href="#module-sonia.sequence_generation" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="sonia.sequence_generation.SequenceGeneration">
<em class="property">class </em><code class="sig-prename descclassname">sonia.sequence_generation.</code><code class="sig-name descname">SequenceGeneration</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sonia_model</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">custom_olga_model</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">custom_genomic_data</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sequence_generation.SequenceGeneration" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class used to evaluate sequences with the sonia model</p>
<dl class="simple">
<dt>sonia_model: object </dt><dd><p>Required. Sonia model: only object accepted.</p>
</dd>
<dt>custom_olga_model: object </dt><dd><p>Optional: already loaded custom olga sequence_generation object</p>
</dd>
<dt>custom_genomic_data: object</dt><dd><p>Optional: already loaded custom olga genomic_data object</p>
</dd>
</dl>
<dl class="simple">
<dt>generate_sequences_pre(num_seqs = 1)</dt><dd><p>Generate sequences using olga</p>
</dd>
<dt>generate_sequences_post(num_seqs,upper_bound=10)</dt><dd><p>Generate sequences using olga and perform rejection selection.</p>
</dd>
<dt>rejection_sampling(upper_bound=10,energies=None)</dt><dd><p>Returns acceptance from rejection sampling of a list of energies.
By default uses the generated sequences within the sonia model.</p>
</dd>
</dl>
<dl class="py method">
<dt id="sonia.sequence_generation.SequenceGeneration.generate_sequences_post">
<code class="sig-name descname">generate_sequences_post</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_seqs</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">upper_bound</span><span class="o">=</span><span class="default_value">10</span></em>, <em class="sig-param"><span class="n">nucleotide</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sequence_generation.SequenceGeneration.generate_sequences_post" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates MonteCarlo sequences from Sonia through rejection sampling.</p>
<dl class="simple">
<dt>num_seqs<span class="classifier">int or float</span></dt><dd><p>Number of MonteCarlo sequences to generate and add to the specified
sequence pool.</p>
</dd>
<dt>upper_bound: int</dt><dd><p>accept all above the threshold. Relates to the percentage of 
sequences that pass selection.</p>
</dd>
</dl>
<dl class="simple">
<dt>seqs<span class="classifier">list</span></dt><dd><p>MonteCarlo sequences drawn from a VDJ recomb model that pass selection.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sequence_generation.SequenceGeneration.generate_sequences_pre">
<code class="sig-name descname">generate_sequences_pre</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_seqs</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">nucleotide</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sequence_generation.SequenceGeneration.generate_sequences_pre" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates MonteCarlo sequences for gen_seqs using OLGA.</p>
<p>Only generates seqs from a V(D)J model. Requires the OLGA package
(pip install olga).</p>
<dl class="simple">
<dt>num_seqs<span class="classifier">int or float</span></dt><dd><p>Number of MonteCarlo sequences to generate and add to the specified
sequence pool.</p>
</dd>
</dl>
<dl class="simple">
<dt>seqs<span class="classifier">list</span></dt><dd><p>MonteCarlo sequences drawn from a VDJ recomb model</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sequence_generation.SequenceGeneration.rejection_sampling">
<code class="sig-name descname">rejection_sampling</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">upper_bound</span><span class="o">=</span><span class="default_value">10</span></em>, <em class="sig-param"><span class="n">energies</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sequence_generation.SequenceGeneration.rejection_sampling" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns acceptance from rejection sampling of a list of seqs.
By default uses the generated sequences within the sonia model.</p>
<dl class="simple">
<dt>upper_bound<span class="classifier">int or float</span></dt><dd><p>accept all above the threshold. Relates to the percentage of 
sequences that pass selection</p>
</dd>
</dl>
<dl class="simple">
<dt>rejection selection: array of bool</dt><dd><p>acceptance of each sequence.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-sonia.sonia">
<span id="sonia-sonia-module"></span><h2>sonia.sonia module<a class="headerlink" href="#module-sonia.sonia" title="Permalink to this headline">¶</a></h2>
<p>Created on Wed Jan 30 12:06:58 2019</p>
<p>&#64;author: zacharysethna</p>
<dl class="py class">
<dt id="sonia.sonia.Sonia">
<em class="property">class </em><code class="sig-prename descclassname">sonia.sonia.</code><code class="sig-name descname">Sonia</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">features</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">data_seqs</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">gen_seqs</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">chain_type</span><span class="o">=</span><span class="default_value">'humanTRB'</span></em>, <em class="sig-param"><span class="n">load_dir</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">feature_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">model_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">data_seq_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gen_seq_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">log_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">load_seqs</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">l2_reg</span><span class="o">=</span><span class="default_value">0.0</span></em>, <em class="sig-param"><span class="n">min_energy_clip</span><span class="o">=</span><span class="default_value">- 5</span></em>, <em class="sig-param"><span class="n">max_energy_clip</span><span class="o">=</span><span class="default_value">10</span></em>, <em class="sig-param"><span class="n">seed</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">vj</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia.Sonia" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class used to infer a Q selection model.</p>
<dl class="simple">
<dt>features<span class="classifier">ndarray</span></dt><dd><dl class="simple">
<dt>Array of feature lists. Each list contains individual subfeatures which</dt><dd><p>all must be satisfied.</p>
</dd>
</dl>
</dd>
<dt>features_dict<span class="classifier">dict</span></dt><dd><p>Dictionary keyed by tuples of the feature lists. Values are the index
of the feature, i.e. self.features[self.features_dict[tuple(f)]] = f.</p>
</dd>
<dt>constant_features<span class="classifier">list</span></dt><dd><dl class="simple">
<dt>List of feature strings to not update parameters during learning. These</dt><dd><p>features are still used to compute energies (not currently used)</p>
</dd>
</dl>
</dd>
<dt>data_seqs<span class="classifier">list</span></dt><dd><p>Data sequences used to infer selection model. Note, each ‘sequence’
is a list where the first element is the CDR3 sequence which is
followed by any V or J genes.</p>
</dd>
<dt>gen_seqs<span class="classifier">list</span></dt><dd><p>Sequences from generative distribution used to infer selection model.
Note, each ‘sequence’ is a list where the first element is the CDR3
sequence which is followed by any V or J genes.</p>
</dd>
<dt>data_seq_features<span class="classifier">list</span></dt><dd><p>Lists of features that data_seqs project onto</p>
</dd>
<dt>gen_seq_features<span class="classifier">list</span></dt><dd><p>Lists of features that gen_seqs project onto</p>
</dd>
<dt>data_marginals<span class="classifier">ndarray</span></dt><dd><p>Array of the marginals of each feature over data_seqs</p>
</dd>
<dt>gen_marginals<span class="classifier">ndarray</span></dt><dd><p>Array of the marginals of each feature over gen_seqs</p>
</dd>
<dt>model_marginals<span class="classifier">ndarray</span></dt><dd><p>Array of the marginals of each feature over the model weighted gen_seqs</p>
</dd>
<dt>L1_converge_history<span class="classifier">list</span></dt><dd><p>L1 distance between data_marginals and model_marginals at each
iteration.</p>
</dd>
<dt>chain_type<span class="classifier">str</span></dt><dd><p>Type of receptor. This specification is used to determine gene names
and allow integrated OLGA sequence generation. Options: ‘humanTRA’,
‘humanTRB’ (default), ‘humanIGH’, ‘humanIGL’, ‘humanIGK’ and ‘mouseTRB’.</p>
</dd>
<dt>l2_reg<span class="classifier">float or None</span></dt><dd><p>L2 regularization. If None (default) then no regularization.</p>
</dd>
</dl>
<dl class="simple">
<dt>seq_feature_proj(feature, seq)</dt><dd><p>Determines if a feature matches/is found in a sequence.</p>
</dd>
<dt>find_seq_features(seq, features = None)</dt><dd><p>Determines all model features of a sequence.</p>
</dd>
<dt>compute_seq_energy(seq_features = None, seq = None)</dt><dd><p>Computes the energy, as determined by the model, of a sequence.</p>
</dd>
<dt>compute_energy(seqs_features)</dt><dd><p>Computes the energies of a list of seq_features according to the model.</p>
</dd>
<dt>compute_marginals(self, features = None, seq_model_features = None, seqs = None, use_flat_distribution = False)</dt><dd><p>Computes the marginals of features over a set of sequences.</p>
</dd>
<dt>infer_selection(self, epochs = 20, batch_size=5000, initialize = True, seed = None)</dt><dd><p>Infers model parameters (energies for each feature).</p>
</dd>
<dt>update_model_structure(self,output_layer=[],input_layer=[],initialize=False)</dt><dd><p>Sets keras model structure and compiles.</p>
</dd>
<dt>update_model(self, add_data_seqs = [], add_gen_seqs = [], add_features = [], remove_features = [], add_constant_features = [], auto_update_marginals = False, auto_update_seq_features = False)</dt><dd><p>Updates model by adding/removing model features or data/generated seqs.
Marginals and seq_features can also be updated.</p>
</dd>
<dt>add_generated_seqs(self, num_gen_seqs = 0, reset_gen_seqs = True)</dt><dd><p>Generates synthetic sequences using OLGA and adds them to gen_seqs.</p>
</dd>
<dt>plot_model_learning(self, save_name = None)</dt><dd><p>Plots current marginal scatter plot as well as L1 convergence history.</p>
</dd>
<dt>save_model(self, save_dir, attributes_to_save = None)</dt><dd><p>Saves the model.</p>
</dd>
<dt>load_model(self, load_dir, load_seqs = True)</dt><dd><p>Loads a model.</p>
</dd>
</dl>
<dl class="py method">
<dt id="sonia.sonia.Sonia.add_generated_seqs">
<code class="sig-name descname">add_generated_seqs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_gen_seqs</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">reset_gen_seqs</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">custom_model_folder</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">add_error</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">custom_error</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia.Sonia.add_generated_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates MonteCarlo sequences for gen_seqs using OLGA.</p>
<p>Only generates seqs from a V(D)J model. Requires the OLGA package
(pip install olga).</p>
<dl class="simple">
<dt>num_gen_seqs<span class="classifier">int or float</span></dt><dd><p>Number of MonteCarlo sequences to generate and add to the specified
sequence pool.</p>
</dd>
<dt>custom_model_folder<span class="classifier">str</span></dt><dd><p>Path to a folder specifying a custom IGoR formatted model to be
used as a generative model. Folder must contain ‘model_params.txt’
and ‘model_marginals.txt’</p>
</dd>
<dt>add_error: bool</dt><dd><p>simualate sequencing error: default is false</p>
</dd>
<dt>custom_error: int</dt><dd><p>set custom error rate for sequencing error.
Default is the one inferred by igor.</p>
</dd>
</dl>
<dl class="simple">
<dt>gen_seqs<span class="classifier">list</span></dt><dd><p>MonteCarlo sequences drawn from a VDJ recomb model</p>
</dd>
<dt>gen_seq_features<span class="classifier">list</span></dt><dd><p>Features gen_seqs have been projected onto.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia.Sonia.compute_energy">
<code class="sig-name descname">compute_energy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seqs_features</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia.Sonia.compute_energy" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the energy of a list of sequences according to the model.</p>
<dl class="simple">
<dt>seqs_features<span class="classifier">list</span></dt><dd><p>list of encoded sequences into sonia features.</p>
</dd>
</dl>
<dl class="simple">
<dt>E<span class="classifier">float</span></dt><dd><p>Energies of seqs according to the model.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia.Sonia.compute_marginals">
<code class="sig-name descname">compute_marginals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">features</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">seq_model_features</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">seqs</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">use_flat_distribution</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">output_dict</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia.Sonia.compute_marginals" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the marginals of each feature over sequences.
Computes marginals either with a flat distribution over the sequences
or weighted by the model energies. Note, finding the features of each
sequence takes time and should be avoided if it has already been done.
If computing marginals of model features use the default setting to
prevent searching for the model features a second time. Similarly, if
seq_model_features has already been determined use this to avoid
recalculating it.</p>
<dl class="simple">
<dt>features<span class="classifier">list or None</span></dt><dd><p>List of features. This does not need to match the model
features. If None (default) the model features will be used.</p>
</dd>
<dt>seq_features_all<span class="classifier">list</span></dt><dd><p>Indices of model features seqs project onto.</p>
</dd>
<dt>seqs<span class="classifier">list</span></dt><dd><p>List of sequences to compute the feature marginals over. Note, each
‘sequence’ is a list where the first element is the CDR3 sequence
which is followed by any V or J genes.</p>
</dd>
<dt>use_flat_distribution<span class="classifier">bool</span></dt><dd><p>Marginals will be computed using a flat distribution (each seq is
weighted as 1) if True. If False, the marginals are computed using
model weights (each sequence is weighted as exp(-E) = Q). Default
is False.</p>
</dd>
</dl>
<dl class="simple">
<dt>marginals<span class="classifier">ndarray or dict</span></dt><dd><p>Marginals of model features over seqs.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia.Sonia.compute_seq_energy">
<code class="sig-name descname">compute_seq_energy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">seq_features</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia.Sonia.compute_seq_energy" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the energy of a sequence according to the model.</p>
<dl class="simple">
<dt>seq<span class="classifier">list</span></dt><dd><p>CDR3 sequence and any associated genes</p>
</dd>
<dt>seq_features<span class="classifier">list</span></dt><dd><p>Features indices seq projects onto.</p>
</dd>
</dl>
<dl class="simple">
<dt>E<span class="classifier">float</span></dt><dd><p>Energy of seq according to the model.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia.Sonia.find_seq_features">
<code class="sig-name descname">find_seq_features</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq</span></em>, <em class="sig-param"><span class="n">features</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia.Sonia.find_seq_features" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds which features match seq</p>
<dl class="simple">
<dt>seq<span class="classifier">list</span></dt><dd><p>CDR3 sequence and any associated genes</p>
</dd>
<dt>features<span class="classifier">ndarray</span></dt><dd><p>Array of feature lists. Each list contains individual subfeatures which
all must be satisfied.</p>
</dd>
</dl>
<dl class="simple">
<dt>seq_features<span class="classifier">list</span></dt><dd><p>Indices of features seq projects onto.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia.Sonia.infer_selection">
<code class="sig-name descname">infer_selection</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">epochs</span><span class="o">=</span><span class="default_value">10</span></em>, <em class="sig-param"><span class="n">batch_size</span><span class="o">=</span><span class="default_value">5000</span></em>, <em class="sig-param"><span class="n">initialize</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">seed</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">validation_split</span><span class="o">=</span><span class="default_value">0.2</span></em>, <em class="sig-param"><span class="n">monitor</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia.Sonia.infer_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Infer model parameters, i.e. energies for each model feature.</p>
<dl class="simple">
<dt>epochs<span class="classifier">int</span></dt><dd><p>Maximum number of learning epochs</p>
</dd>
<dt>intialize<span class="classifier">bool</span></dt><dd><p>Resets data shuffle</p>
</dd>
<dt>batch_size<span class="classifier">int</span></dt><dd><p>Size of the batches in the inference</p>
</dd>
<dt>seed<span class="classifier">int</span></dt><dd><p>Sets random seed</p>
</dd>
</dl>
<dl class="simple">
<dt>model<span class="classifier">keras model</span></dt><dd><p>Parameters of the model</p>
</dd>
<dt>model_marginals<span class="classifier">array</span></dt><dd><p>Marginals over the generated sequences, reweighted by the model.</p>
</dd>
<dt>L1_converge_history<span class="classifier">list</span></dt><dd><p>L1 distance between data_marginals and model_marginals at each
iteration.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia.Sonia.load_model">
<code class="sig-name descname">load_model</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">load_dir</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">load_seqs</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">feature_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">model_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">data_seq_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gen_seq_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">log_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia.Sonia.load_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads model from directory.</p>
<dl class="simple">
<dt>load_dir<span class="classifier">str</span></dt><dd><p>Directory name to load model attributes from.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia.Sonia.save_model">
<code class="sig-name descname">save_model</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">save_dir</span></em>, <em class="sig-param"><span class="n">attributes_to_save</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">force</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia.Sonia.save_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Saves model parameters and sequences</p>
<dl class="simple">
<dt>save_dir<span class="classifier">str</span></dt><dd><p>Directory name to save model attributes to.</p>
</dd>
<dt>attributes_to_save: list</dt><dd><p>name of attributes to save</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia.Sonia.seq_feature_proj">
<code class="sig-name descname">seq_feature_proj</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">feature</span></em>, <em class="sig-param"><span class="n">seq</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia.Sonia.seq_feature_proj" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks if a sequence matches all subfeatures of the feature list</p>
<dl class="simple">
<dt>feature<span class="classifier">list</span></dt><dd><p>List of individual subfeatures the sequence must match</p>
</dd>
<dt>seq<span class="classifier">list</span></dt><dd><p>CDR3 sequence and any associated genes</p>
</dd>
</dl>
<dl class="simple">
<dt>bool</dt><dd><p>True if seq matches feature else False.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia.Sonia.update_model">
<code class="sig-name descname">update_model</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">add_data_seqs</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">add_gen_seqs</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">add_features</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">remove_features</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">add_constant_features</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">auto_update_marginals</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">auto_update_seq_features</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia.Sonia.update_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Updates the model attributes
This method is used to add/remove model features or data/generated
sequences. These changes will be propagated through the class to update
any other attributes that need to match (e.g. the marginals or
seq_features).</p>
<dl class="simple">
<dt>add_data_seqs<span class="classifier">list</span></dt><dd><p>List of CDR3 sequences to add to data_seq pool.</p>
</dd>
<dt>add_gen_seqs<span class="classifier">list</span></dt><dd><p>List of CDR3 sequences to add to data_seq pool.</p>
</dd>
<dt>add_gen_seqs<span class="classifier">list</span></dt><dd><p>List of CDR3 sequences to add to data_seq pool.</p>
</dd>
<dt>add_features<span class="classifier">list</span></dt><dd><p>List of feature lists to add to self.features</p>
</dd>
<dt>remove_featurese<span class="classifier">list</span></dt><dd><p>List of feature lists and/or indices to remove from self.features</p>
</dd>
<dt>add_constant_features<span class="classifier">list</span></dt><dd><p>List of feature lists to add to constant features. (Not currently used)</p>
</dd>
<dt>auto_update_marginals<span class="classifier">bool</span></dt><dd><p>Specifies to update marginals.</p>
</dd>
<dt>auto_update_seq_features<span class="classifier">bool</span></dt><dd><p>Specifies to update seq features.</p>
</dd>
</dl>
<dl class="simple">
<dt>features<span class="classifier">list</span></dt><dd><p>List of model features</p>
</dd>
<dt>data_seq_features<span class="classifier">list</span></dt><dd><p>Features data_seqs have been projected onto.</p>
</dd>
<dt>gen_seq_features<span class="classifier">list</span></dt><dd><p>Features gen_seqs have been projected onto.</p>
</dd>
<dt>data_marginals<span class="classifier">ndarray</span></dt><dd><p>Marginals over the data sequences for each model feature.</p>
</dd>
<dt>gen_marginals<span class="classifier">ndarray</span></dt><dd><p>Marginals over the generated sequences for each model feature.</p>
</dd>
<dt>model_marginals<span class="classifier">ndarray</span></dt><dd><p>Marginals over the generated sequences, reweighted by the model,
for each model feature.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia.Sonia.update_model_structure">
<code class="sig-name descname">update_model_structure</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">output_layer</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">input_layer</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">initialize</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia.Sonia.update_model_structure" title="Permalink to this definition">¶</a></dt>
<dd><p>Defines the model structure and compiles it.</p>
<dl class="simple">
<dt>structure<span class="classifier">Sequential Model Keras</span></dt><dd><p>structure of the model</p>
</dd>
<dt>initialize: bool</dt><dd><p>if True, it initializes to linear model, otherwise it updates to new structure</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-sonia.sonia_leftpos_rightpos">
<span id="sonia-sonia-leftpos-rightpos-module"></span><h2>sonia.sonia_leftpos_rightpos module<a class="headerlink" href="#module-sonia.sonia_leftpos_rightpos" title="Permalink to this headline">¶</a></h2>
<p>&#64;author: zacharysethna</p>
<dl class="py class">
<dt id="sonia.sonia_leftpos_rightpos.SoniaLeftposRightpos">
<em class="property">class </em><code class="sig-prename descclassname">sonia.sonia_leftpos_rightpos.</code><code class="sig-name descname">SoniaLeftposRightpos</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data_seqs</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">gen_seqs</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">chain_type</span><span class="o">=</span><span class="default_value">'humanTRB'</span></em>, <em class="sig-param"><span class="n">load_dir</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">feature_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">data_seq_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gen_seq_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">log_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">load_seqs</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">max_depth</span><span class="o">=</span><span class="default_value">25</span></em>, <em class="sig-param"><span class="n">max_L</span><span class="o">=</span><span class="default_value">30</span></em>, <em class="sig-param"><span class="n">include_indep_genes</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">include_joint_genes</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">min_energy_clip</span><span class="o">=</span><span class="default_value">- 5</span></em>, <em class="sig-param"><span class="n">max_energy_clip</span><span class="o">=</span><span class="default_value">10</span></em>, <em class="sig-param"><span class="n">seed</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">custom_pgen_model</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">l2_reg</span><span class="o">=</span><span class="default_value">0.0</span></em>, <em class="sig-param"><span class="n">vj</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_leftpos_rightpos.SoniaLeftposRightpos" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#sonia.sonia.Sonia" title="sonia.sonia.Sonia"><code class="xref py py-class docutils literal notranslate"><span class="pre">sonia.sonia.Sonia</span></code></a></p>
<dl class="py method">
<dt id="sonia.sonia_leftpos_rightpos.SoniaLeftposRightpos.add_features">
<code class="sig-name descname">add_features</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">include_indep_genes</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">include_joint_genes</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">custom_pgen_model</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_leftpos_rightpos.SoniaLeftposRightpos.add_features" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a list of feature_lsts for L/R pos model.</p>
<dl class="simple">
<dt>include_genes<span class="classifier">bool</span></dt><dd><p>If true, features for gene selection are also generated. Currently
joint V/J pairs used.</p>
</dd>
<dt>custom_pgen_model: string</dt><dd><p>path to folder of custom olga model.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia_leftpos_rightpos.SoniaLeftposRightpos.compute_seq_energy_from_parameters">
<code class="sig-name descname">compute_seq_energy_from_parameters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seqs</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">seqs_features</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_leftpos_rightpos.SoniaLeftposRightpos.compute_seq_energy_from_parameters" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the energy of a list of sequences according to the model.</p>
<p>This computes according to model parameters instead of the keras model.
As a result, no clipping occurs.</p>
<dl class="simple">
<dt>seqs<span class="classifier">list or None</span></dt><dd><p>Sequence list for a single sequence or many.</p>
</dd>
<dt>seqs_features<span class="classifier">list or None</span></dt><dd><p>list of sequence features for a single sequence or many.</p>
</dd>
</dl>
<dl class="simple">
<dt>E<span class="classifier">float</span></dt><dd><p>Energies of seqs according to the model.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia_leftpos_rightpos.SoniaLeftposRightpos.find_seq_features">
<code class="sig-name descname">find_seq_features</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq</span></em>, <em class="sig-param"><span class="n">features</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_leftpos_rightpos.SoniaLeftposRightpos.find_seq_features" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds which features match seq</p>
<p>If no features are provided, the left/right indexing amino acid model
features will be assumed.</p>
<dl class="simple">
<dt>seq<span class="classifier">list</span></dt><dd><p>CDR3 sequence and any associated genes</p>
</dd>
<dt>features<span class="classifier">ndarray</span></dt><dd><p>Array of feature lists. Each list contains individual subfeatures which
all must be satisfied.</p>
</dd>
</dl>
<dl class="simple">
<dt>seq_features<span class="classifier">list</span></dt><dd><p>Indices of features seq projects onto.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia_leftpos_rightpos.SoniaLeftposRightpos.get_energy_parameters">
<code class="sig-name descname">get_energy_parameters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">return_as_dict</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_leftpos_rightpos.SoniaLeftposRightpos.get_energy_parameters" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract energy terms from keras model.</p>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia_leftpos_rightpos.SoniaLeftposRightpos.save_model">
<code class="sig-name descname">save_model</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">save_dir</span></em>, <em class="sig-param"><span class="n">attributes_to_save</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">force</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_leftpos_rightpos.SoniaLeftposRightpos.save_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Saves model parameters and sequences</p>
<dl class="simple">
<dt>save_dir<span class="classifier">str</span></dt><dd><p>Directory name to save model attributes to.</p>
</dd>
<dt>attributes_to_save: list</dt><dd><p>Names of attributes to save</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-sonia.sonia_length_pos">
<span id="sonia-sonia-length-pos-module"></span><h2>sonia.sonia_length_pos module<a class="headerlink" href="#module-sonia.sonia_length_pos" title="Permalink to this headline">¶</a></h2>
<p>Created on Wed Mar  6 15:12:15 2019</p>
<p>&#64;author: administrator</p>
<dl class="py class">
<dt id="sonia.sonia_length_pos.SoniaLengthPos">
<em class="property">class </em><code class="sig-prename descclassname">sonia.sonia_length_pos.</code><code class="sig-name descname">SoniaLengthPos</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data_seqs</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">gen_seqs</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">chain_type</span><span class="o">=</span><span class="default_value">'humanTRB'</span></em>, <em class="sig-param"><span class="n">load_dir</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">feature_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">data_seq_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gen_seq_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">log_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">min_L</span><span class="o">=</span><span class="default_value">4</span></em>, <em class="sig-param"><span class="n">max_L</span><span class="o">=</span><span class="default_value">30</span></em>, <em class="sig-param"><span class="n">include_indep_genes</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">include_joint_genes</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">min_energy_clip</span><span class="o">=</span><span class="default_value">- 5</span></em>, <em class="sig-param"><span class="n">max_energy_clip</span><span class="o">=</span><span class="default_value">10</span></em>, <em class="sig-param"><span class="n">seed</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">custom_pgen_model</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">l2_reg</span><span class="o">=</span><span class="default_value">0.0</span></em>, <em class="sig-param"><span class="n">vj</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_length_pos.SoniaLengthPos" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#sonia.sonia.Sonia" title="sonia.sonia.Sonia"><code class="xref py py-class docutils literal notranslate"><span class="pre">sonia.sonia.Sonia</span></code></a></p>
<dl class="py method">
<dt id="sonia.sonia_length_pos.SoniaLengthPos.add_features">
<code class="sig-name descname">add_features</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">include_indep_genes</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">include_joint_genes</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">custom_pgen_model</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_length_pos.SoniaLengthPos.add_features" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a list of feature_lsts for a length dependent L pos model.</p>
<dl class="simple">
<dt>include_genes<span class="classifier">bool</span></dt><dd><p>If true, features for gene selection are also generated. Currently
joint V/J pairs used.</p>
</dd>
<dt>custom_pgen_model: string</dt><dd><p>path to folder of custom olga model.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia_length_pos.SoniaLengthPos.compute_seq_energy_from_parameters">
<code class="sig-name descname">compute_seq_energy_from_parameters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seqs</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">seqs_features</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_length_pos.SoniaLengthPos.compute_seq_energy_from_parameters" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the energy of a list of sequences according to the model.</p>
<p>This computes according to model parameters instead of the keras model.
As a result, no clipping occurs.</p>
<dl class="simple">
<dt>seqs<span class="classifier">list or None</span></dt><dd><p>Sequence list for a single sequence or many.</p>
</dd>
<dt>seqs_features<span class="classifier">list or None</span></dt><dd><p>list of sequence features for a single sequence or many.</p>
</dd>
</dl>
<dl class="simple">
<dt>E<span class="classifier">float</span></dt><dd><p>Energies of seqs according to the model.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia_length_pos.SoniaLengthPos.find_seq_features">
<code class="sig-name descname">find_seq_features</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq</span></em>, <em class="sig-param"><span class="n">features</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_length_pos.SoniaLengthPos.find_seq_features" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds which features match seq</p>
<p>If no features are provided, the length dependent amino acid model
features will be assumed.</p>
<dl class="simple">
<dt>seq<span class="classifier">list</span></dt><dd><p>CDR3 sequence and any associated genes</p>
</dd>
<dt>features<span class="classifier">ndarray</span></dt><dd><p>Array of feature lists. Each list contains individual subfeatures which
all must be satisfied.</p>
</dd>
</dl>
<dl class="simple">
<dt>seq_features<span class="classifier">list</span></dt><dd><p>Indices of features seq projects onto.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia_length_pos.SoniaLengthPos.get_energy_parameters">
<code class="sig-name descname">get_energy_parameters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">return_as_dict</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_length_pos.SoniaLengthPos.get_energy_parameters" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract energy terms from keras model and gauge.</p>
<p>For the length dependent position model, the gauge is set so that at a
given position, for a given length, we have:</p>
<p>&lt;q_i,aa;L&gt;_gen|L = 1</p>
<dl class="simple">
<dt>min_L<span class="classifier">int</span></dt><dd><p>Minimum length CDR3 sequence, if not given taken from class attribute</p>
</dd>
<dt>max_L<span class="classifier">int</span></dt><dd><p>Maximum length CDR3 sequence, if not given taken from class attribute</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia_length_pos.SoniaLengthPos.save_model">
<code class="sig-name descname">save_model</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">save_dir</span></em>, <em class="sig-param"><span class="n">attributes_to_save</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">force</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_length_pos.SoniaLengthPos.save_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Saves model parameters and sequences</p>
<dl class="simple">
<dt>save_dir<span class="classifier">str</span></dt><dd><p>Directory name to save model attributes to.</p>
</dd>
<dt>attributes_to_save: list</dt><dd><p>Names of attributes to save</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-sonia.sonia_vjl">
<span id="sonia-sonia-vjl-module"></span><h2>sonia.sonia_vjl module<a class="headerlink" href="#module-sonia.sonia_vjl" title="Permalink to this headline">¶</a></h2>
<p>&#64;author: Giulio Isacchini</p>
<dl class="py class">
<dt id="sonia.sonia_vjl.SoniaVJL">
<em class="property">class </em><code class="sig-prename descclassname">sonia.sonia_vjl.</code><code class="sig-name descname">SoniaVJL</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data_seqs</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">gen_seqs</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">chain_type</span><span class="o">=</span><span class="default_value">'humanTRB'</span></em>, <em class="sig-param"><span class="n">load_dir</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">feature_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">data_seq_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gen_seq_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">log_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">load_seqs</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">max_depth</span><span class="o">=</span><span class="default_value">25</span></em>, <em class="sig-param"><span class="n">max_L</span><span class="o">=</span><span class="default_value">30</span></em>, <em class="sig-param"><span class="n">include_indep_genes</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">include_joint_genes</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">min_energy_clip</span><span class="o">=</span><span class="default_value">- 5</span></em>, <em class="sig-param"><span class="n">max_energy_clip</span><span class="o">=</span><span class="default_value">10</span></em>, <em class="sig-param"><span class="n">seed</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">custom_pgen_model</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">l2_reg</span><span class="o">=</span><span class="default_value">0.0</span></em>, <em class="sig-param"><span class="n">vj</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">joint_vjl</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_vjl.SoniaVJL" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#sonia.sonia.Sonia" title="sonia.sonia.Sonia"><code class="xref py py-class docutils literal notranslate"><span class="pre">sonia.sonia.Sonia</span></code></a></p>
<dl class="py method">
<dt id="sonia.sonia_vjl.SoniaVJL.add_features">
<code class="sig-name descname">add_features</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">custom_pgen_model</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_vjl.SoniaVJL.add_features" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a list of feature_lsts for L/R pos model.</p>
<dl class="simple">
<dt>include_genes<span class="classifier">bool</span></dt><dd><p>If true, features for gene selection are also generated. Currently
joint V/J pairs used.</p>
</dd>
<dt>custom_pgen_model: string</dt><dd><p>path to folder of custom olga model.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia_vjl.SoniaVJL.compute_seq_energy_from_parameters">
<code class="sig-name descname">compute_seq_energy_from_parameters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seqs</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">seqs_features</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_vjl.SoniaVJL.compute_seq_energy_from_parameters" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the energy of a list of sequences according to the model.</p>
<p>This computes according to model parameters instead of the keras model.
As a result, no clipping occurs.</p>
<dl class="simple">
<dt>seqs<span class="classifier">list or None</span></dt><dd><p>Sequence list for a single sequence or many.</p>
</dd>
<dt>seqs_features<span class="classifier">list or None</span></dt><dd><p>list of sequence features for a single sequence or many.</p>
</dd>
</dl>
<dl class="simple">
<dt>E<span class="classifier">float</span></dt><dd><p>Energies of seqs according to the model.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia_vjl.SoniaVJL.find_seq_features">
<code class="sig-name descname">find_seq_features</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq</span></em>, <em class="sig-param"><span class="n">features</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_vjl.SoniaVJL.find_seq_features" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds which features match seq</p>
<p>If no features are provided, the left/right indexing amino acid model
features will be assumed.</p>
<dl class="simple">
<dt>seq<span class="classifier">list</span></dt><dd><p>CDR3 sequence and any associated genes</p>
</dd>
<dt>features<span class="classifier">ndarray</span></dt><dd><p>Array of feature lists. Each list contains individual subfeatures which
all must be satisfied.</p>
</dd>
</dl>
<dl class="simple">
<dt>seq_features<span class="classifier">list</span></dt><dd><p>Indices of features seq projects onto.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia_vjl.SoniaVJL.get_energy_parameters">
<code class="sig-name descname">get_energy_parameters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">return_as_dict</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_vjl.SoniaVJL.get_energy_parameters" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract energy terms from keras model.</p>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia_vjl.SoniaVJL.infer_selection">
<code class="sig-name descname">infer_selection</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">epochs</span><span class="o">=</span><span class="default_value">10</span></em>, <em class="sig-param"><span class="n">batch_size</span><span class="o">=</span><span class="default_value">5000</span></em>, <em class="sig-param"><span class="n">initialize</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">seed</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">validation_split</span><span class="o">=</span><span class="default_value">0.2</span></em>, <em class="sig-param"><span class="n">monitor</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_vjl.SoniaVJL.infer_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Infer model parameters, i.e. energies for each model feature.</p>
<dl class="simple">
<dt>epochs<span class="classifier">int</span></dt><dd><p>Maximum number of learning epochs</p>
</dd>
<dt>intialize<span class="classifier">bool</span></dt><dd><p>Resets data shuffle</p>
</dd>
<dt>batch_size<span class="classifier">int</span></dt><dd><p>Size of the batches in the inference</p>
</dd>
<dt>seed<span class="classifier">int</span></dt><dd><p>Sets random seed</p>
</dd>
</dl>
<dl class="simple">
<dt>model<span class="classifier">keras model</span></dt><dd><p>Parameters of the model</p>
</dd>
<dt>model_marginals<span class="classifier">array</span></dt><dd><p>Marginals over the generated sequences, reweighted by the model.</p>
</dd>
<dt>L1_converge_history<span class="classifier">list</span></dt><dd><p>L1 distance between data_marginals and model_marginals at each
iteration.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sonia.sonia_vjl.SoniaVJL.save_model">
<code class="sig-name descname">save_model</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">save_dir</span></em>, <em class="sig-param"><span class="n">attributes_to_save</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sonia.sonia_vjl.SoniaVJL.save_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Saves model parameters and sequences</p>
<dl class="simple">
<dt>save_dir<span class="classifier">str</span></dt><dd><p>Directory name to save model attributes to.</p>
</dd>
<dt>attributes_to_save: list</dt><dd><p>Names of attributes to save</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="license.html" class="btn btn-neutral float-right" title="License" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="modules.html" class="btn btn-neutral float-left" title="Documentation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Zachary Sethna and Giulio Isacchini

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>